### DISCUSSION TITLE
time_pulse_us triggered 5 times after timeout. Why?

### DISCUSSION THREAD

**Discussion by sk8board** [micropython/micropython#16045] [Using MicroPython] [answered]
I am using the `time_pulse_us()` function to detect when a signal has stopped for more than 2 seconds. This signal is not transmitted often, but I need to trigger another function 2 seconds after the signal stops. The use of `time_pulse_us` works as expected except when a timeout occurs.

When a timeout occurs, `time_pulse_us` will execute the entire function that contains `time_pulse_us`. When the timeout occurs again, then the entire function gets executed again. This repeats five times for a total of 10 seconds, which blocks other code from running.

Why does `time_pulse_us` execute an entire function 5 times after timeout? 

What would be a better way to detect when a digital signal times-out which does not block or consume much processor?

``` py
def sense2_signal(x):
    if (machine.time_pulse_us(sense2,1,2000000)) == -1:
        Yellow()


sense2.irq(sense2_signal,machine.Pin.IRQ_RISING)
```

**Comment by robert-hh** [comment, marked as answer]
You did not tell which port you use.

In your code you set up a Pin irq to call sense2_signal(). It is a soft irq, so the Pin.irq is sheduled for every rising slope that occurs and sense2_signal may be called several times. If the input signal can have several fast transitions like they are created from a bouncing switch, one can get a burst of IRQ calls.
Also, you mentioned it, having the IRQ handler to block for 2 seconds is not good.

I would set up two irq handlers, the Pin IRQ for detecting input transitions and a timer irq to detect the timeout, which just measures the time between a start time and the actual time. The Pin irq would reset the start time to the actual time. Time sampling can be done with time.ticks_ms() and time.ticks_diff(). The frequency of the timer IRQ is up to the required precision, from a fee ms up. That way nothing is blocking an Pin irq bursts will not harm.

Edit: And of course you can use asyncio.

**Reply by sk8board** [reply]
> You did not tell which port you use.

Sorry. RP2040

> In your code you set up a Pin irq to call sense2_signal(). It is a soft irq, so the Pin.irq is sheduled for every rising slope that occurs and sense2_signal may be called several times. If the input signal can have several fast transitions like they are created from a bouncing switch, one can get a burst of IRQ calls. Also, you mentioned it, having the IRQ handler to block for 2 seconds is not good.

Good catch. I connected an oscilloscope and found severe signal bounce. 

> I would set up two irq handlers, the Pin IRQ for detecting input transitions and a timer irq to detect the timeout, which just measures the time between a start time and the actual time. The Pin irq would reset the start time to the actual time. Time sampling can be done with time.ticks_ms() and time.ticks_diff(). The frequency of the timer IRQ is up to the required precision, from a fee ms up. That way nothing is blocking an Pin irq bursts will not harm.

Thank you for this suggestion. I tried this but found the interrupt scheduler was overloaded due to the severe signal bounce. I replaced the Pin irq for polling with timer interrupt. This worked much better.

> Edit: And of course you can use asyncio.

Another great suggestion. I was trying to avoid this as I wanted a simpler solution. I will keep this in mind if I need to improve efficiency.

Thanks again!
